alf/utils/checkpoint_utils_test.py

```diff
@@ -0,0 +1,549 @@
+                                               'state': {},
+                                               'param_groups': [{
+                                                   'lr': 0.2,
+                                                   'betas': (0.9, 0.999),
+                                                   'eps': 1e-08,
+                                                   'weight_decay': 0,
+                                                   'amsgrad': False,
+                                                   'params': []
+                                               }]
+                                           }),
+                                           ('_param_list.0',
+                                            torch.tensor([0.])),
+                                           ('_optimizers.0', {
+                                               'state': {},
+                                               'param_groups': [{
+                                                   'lr': 0.1,
+                                                   'betas': (0.9, 0.999),
+                                                   'eps': 1e-08,
+                                                   'weight_decay': 0,
+                                                   'amsgrad': False,
+                                                   'params': []
+                                               }]
+                                           })])
+
+        ckpt_mngr = ckpt_utils.Checkpointer(ckpt_dir, alg=alg_root)
+        ckpt_mngr.save(0)
+
+        # modify some parameter values after saving
+        with torch.no_grad():
+            alg_root._sub_alg1._param_list[0].copy_(torch.Tensor([-1]))
+
+        self.assertTrue((alg_root.state_dict() != expected_state_dict))
+
+        # recover the expected values after loading
+        ckpt_mngr.load(0)
+        self.assertTrue((alg_root.state_dict() == expected_state_dict))
+
+        # case 2: cycle with ignore
+        alg_root2 = ComposedAlgWithIgnore(
+            params=[param_root],
+            optimizer=optimizer_root,
+            sub_alg1=alg_1,
+            sub_alg2=alg_2,
+            name="root")
+
+        alg_2.root = alg_root2
+
+        ckpt_mngr = ckpt_utils.Checkpointer(ckpt_dir, alg=alg_root2)
+        ckpt_mngr.save(0)
+
+        # modify some parameter values after saving
+        with torch.no_grad():
+            alg_root._sub_alg1._param_list[0].copy_(torch.Tensor([-1]))
+
+        self.assertTrue((alg_root.state_dict() != expected_state_dict))
+
+        # recover the expected values after loading
+        ckpt_mngr.load(0)
+        self.assertTrue((alg_root2.state_dict() == expected_state_dict))
+
+
+class TestModelMismatch(unittest.TestCase):
+    def test_model_mismatch(self):
+        # test model mis-match
+
+        ckpt_dir = "/tmp/ckpt_data/model_mis_match/"
+
+        if os.path.isdir(ckpt_dir):
```

Comment:
For all the tests that use tmp files, suggest using `tempfile` context manager that will automatically create and delete tmp directories.

Reply:
good suggestion! changed